umap/umap_.py

```diff
@@ -1631,6 +1777,149 @@ def transform(self, X):
+        distances = [np.array([umap.distances.euclidean(X[i], self.embedding_[nb]) for nb in neighborhood[i]]) for i in range(X.shape[0])]
+        idx = np.array([np.argsort(e)[:min_vertices] for e in distances])
+
+        dists_output_space = np.array([distances[i][idx[i]] for i in range(len(distances))])
+        indices = np.array([neighborhood[i][idx[i]] for i in range(len(neighborhood))])
+
+        rows, cols, distances = np.array([
+            [i, indices[i,j], dists_output_space[i,j]]
+            for i in range(indices.shape[0])
+            for j in range(min_vertices)
+        ]).T
+
+        # calculate membership strength of each edge
+        weights = 1 / (1 + self._a * distances ** (2 * self._b))
+
+        # compute 1-skeleton
+        # convert 1-skeleton into coo_matrix adjacency matrix
+        graph = scipy.sparse.coo_matrix(
+            (weights, (rows, cols)), shape=(X.shape[0], self._raw_data.shape[0])
+        )
+
+        a = np.array(neighborhood)
+        b = 1 / (1 + self._a * np.array(distances) ** (2 * self._b))
+
+        # initialize transformationn with linear combinations of neighboring points
+        inv_transformed_points = init_transform(
+            indices,
+            1 / (1 + self._a * np.array(dists_output_space) ** (2 * self._b)),
+            self._raw_data
+        )
+
+        if self.n_epochs is None:
+            # For smaller datasets we can use more epochs
+            if graph.shape[0] <= 10000:
+                n_epochs = 100
+            else:
+                n_epochs = 30
+        else:
+            n_epochs = self.n_epochs
+
+        graph.data[graph.data < (graph.data.max() / float(n_epochs))] = 0.0
+        graph.eliminate_zeros()
+
+        epochs_per_sample = make_epochs_per_sample(graph.data, n_epochs)
+
+        head = graph.row
+        tail = graph.col
+        weight = graph.data
+
+        if callable(self.input_metric):
+            _input_distance_func = self.input_metric
+        elif self.input_metric in umap.distances.named_distances and self.input_metric in umap.distances.named_distances_with_gradients:
+            _input_distance_func = umap.distances.named_distances_with_gradients[self.input_metric]
+        elif self.input_metric == 'precomputed':
+            raise ValueError(
+                "input_metric cannnot be 'precomputed'"
+            )
+        else:
+            if self.output_metric in umap.distances.named_distances:
+                raise ValueError(
+                    "gradient function is not yet implemented for " + repr(self.output_metric) + "."
+                )
+            else:
+                raise ValueError(
+                    "output_metric is neither callable, " + "nor a recognised string"
+                )
+
+
+        optimize_layout = make_optimize_layout(
+            _input_distance_func,
+            tuple(self._input_metric_kwds.values()),
+        )
+
+        sigmas, rhos = smooth_knn_dist(
```

Comment:
I suspect this is not what you want here, unless the sigmas are playing a different role than I would expect?